src/systems/sprites.js

```diff
@@ -187,224 +177,234 @@ function createGeometry(maxSprites) {
-    this.meshes = {};
-
-    for (const stack of Object.values(this.stack)) {
-      for (let i = 0; i < this.maxSprites; i++) {
-        stack[i] = this.maxSprites - 1 - i;
-      }
-    }
 
-    const vertexShader = String.prototype.concat(
-      scene.renderer.xr.multiview ? multiviewVertPrefix : nonmultiviewVertPrefix,
-      vert
-    );
-    const fragmentShader = String.prototype.concat(
-      scene.renderer.xr.multiview ? multiviewFragPrefix : nonmultiviewFragPrefix,
-      frag
-    );
-
-    const domReady = waitForDOMContentLoaded();
-
-    Promise.all([domReady]).then(() => {
-      for (const [spritesheetPng, type] of pngs) {
-        Promise.all([createImageTexture(spritesheetPng, getThemeColorShifter(type)), waitForDOMContentLoaded()]).then(
-          ([spritesheetTexture]) => {
-            const material = new THREE.RawShaderMaterial({
-              uniforms: {
-                u_spritesheet: { value: spritesheetTexture },
-                hubs_Time: { value: 0 }
-              },
-              vertexShader,
-              fragmentShader,
-              side: THREE.DoubleSide,
-              transparent: true
-            });
-            const mesh = (this.meshes[type] = new THREE.Mesh(createGeometry(this.maxSprites), material));
-            const el = document.createElement("a-entity");
-            el.classList.add("ui");
-            scene.appendChild(el);
-            el.setObject3D("mesh", mesh);
-            mesh.frustumCulled = false;
-            mesh.renderOrder = window.APP.RENDER_ORDER.HUD_ICONS;
-            mesh.raycast = this.raycast.bind(this);
-          }
-        );
+  constructor(scene) {
+    this.sprites = { action: [], notice: [] };
+    this.indicesFromSprites = { action: new WeakMap(), notice: new WeakMap() };
+    this.meshes = { action: null, notice: null };
+
+    waitForDOMContentLoaded().then(() => {
+      for (const type in PNGS) {
+        const spritesheetPng = PNGS[type];
+        createImageTexture(spritesheetPng, getThemeColorShifter(type)).then(spritesheetTexture => {
+          const geometry = createGeometry(MAX_SPRITES);
+          const material = createMaterial(spritesheetTexture);
+          const mesh = (this.meshes[type] = new THREE.Mesh(geometry, material));
+          const el = document.createElement("a-entity");
+          el.classList.add("ui");
+          scene.appendChild(el);
+          el.setObject3D("mesh", mesh);
+          mesh.frustumCulled = false;
+          mesh.renderOrder = window.APP.RENDER_ORDER.HUD_ICONS;
+          mesh.raycast = this.raycast.bind(this);
+        });
       }
     });
 
-    APP.store.addEventListener("themechanged", async () => {
-      for (const [spritesheetPng, type] of pngs) {
+    APP.store.addEventListener("themechanged", () => {
+      for (const type in PNGS) {
+        const spritesheetPng = PNGS[type];
+        // TODO: Fix me if possible
```

Comment:
Yeah. I would eventually like to rework this theming stuff such that sprites are colorless and can individually have a color property that is applied in the shader. We might then also use an SDF for the sprites themselves too, depending on if we end up needing sprites to support multiple tones.